feat: introduce request group settings (name, description, and move/copy)

NickHackman opened the pull request:
Adds the ability to modify the description for a Request Group in a modal shown by hitting settings in a dropdown. Identical implementation to RequestSettings. In addition allows renaming the Request Group as well as moving/copying it to a different workspace.

closes: https://github.com/Kong/insomnia/issues/3306

## Screenshots

![Screen Shot 2021-05-02 at 6 48 11 PM](https://user-images.githubusercontent.com/31719071/116831791-0cd9f280-ab77-11eb-8cee-71537fd958a3.png)

Following pattern set by Requests as this goes after plugins.

![Screen Shot 2021-05-02 at 6 48 34 PM](https://user-images.githubusercontent.com/31719071/116831799-23804980-ab77-11eb-9a62-138cbe0e5832.png)

The words "Request Group" here feels wordy personally, I prefer "Folder", but whatever you think is best. This also adds the support of copying/moving workspaces as it is almost identical to the Request implementation. 

> I think opening the docs (in the request view/area) and expanding the folder/group is an obvious and solid approach. It would also have the benefit of being easily expanded upon, allowing in the future to have Folder Plugins as buttons on the far right side of where requests have headers/authorization or other group/folder features (ie: authorization, headers, etc). Anyways, just like to ensure that I'm on the same page here.

My comment here, I feel like this should be a different issue, but I'm down to address it if you'd like, I personally really want some of those future features as being able to add authorization headers for ALL requests in a group would be AWESOME!

I appreciate all feedback, thanks! 👍

NickHackman committed:
feat: request group descriptions

Adds the ability to modify the description for a Request Group in a modal shown by hitting settings in a dropdown. Identical implementation to RequestSettings. In addition allows renaming the Request Group as well as moving/copying it to a different workspace.

Ref: https://github.com/Kong/insomnia/issues/3306

develohpanda commented on packages/insomnia-app/app/ui/components/sidebar/sidebar-request-group-row.js:
```diff
@@ -47,6 +49,10 @@ class SidebarRequestGroupRow extends PureComponent {
     }
   }
 
+  _handleShowRequestSettings() {
```
```suggestion
  _handleShowRequestGroupSettings() {
```

develohpanda commented on packages/insomnia-app/app/ui/components/sidebar/sidebar-request-group-row.js:
```diff
@@ -123,6 +129,7 @@ class SidebarRequestGroupRow extends PureComponent {
               handleCreateRequest={handleCreateRequest}
               handleCreateRequestGroup={handleCreateRequestGroup}
               handleDuplicateRequestGroup={handleDuplicateRequestGroup}
+              handleShowSettings={this._handleShowRequestSettings}
```
```suggestion
              handleShowSettings={this._handleShowRequestGroupSettings}
```

develohpanda commented on packages/insomnia-app/app/ui/components/modals/request-group-settings-modal.js:
```diff
@@ -0,0 +1,338 @@
+// @flow
```
I diff'd this component against the one it was derived from - `RequestSettingsModal` - the changes are minimal and correct.

My view on this duplication - because `RequestGroup` and `Request` are fundamentally different entities I think it's better to keep their implementations separate for the time being, instead of refactoring out a generic component. In the future should we need something like this _again_ it might be worth exploring a base `EntitySettings` component.

Overall, nice work 👏🏽

develohpanda commented on packages/insomnia-app/app/ui/components/modals/request-group-settings-modal.js:
```diff
@@ -0,0 +1,338 @@
+      <div>
+        <div className="form-control form-control--outlined">
+          <label>
+            Name{' '}
+            <span className="txt-sm faint italic">(also rename by double-clicking in sidebar)</span>
```
Double-clicking a folder just expands or collapses it 😆 

```suggestion
            Name
```

develohpanda reviewed (commented):
Thanks for the PR, and welcome to the project! We currently have a code-freeze in place because we're transitioning to typescript (1:1, should be fairly simple to update this PR), so I won't merge or test this just yet. 🤗 

> The words "Request Group" here feels wordy personally, I prefer "Folder", but whatever you think is best.

Let's go with folder! `RequestGroup` is just the verbiage used in the code but at some point in the past the UI started to use `Folder`.

> This also adds the support of copying/moving workspaces as it is almost identical to the Request implementation.

That seems fine to me because we get it for free by deriving the implementation from `RequestSettingsModal`. It would be worth checking there is nothing special happening in the "Move" dropdown item here. If so, let's bring that over and remove this entry.
<img width="250" alt="image" src="https://user-images.githubusercontent.com/4312346/116860650-ffe00080-ac55-11eb-976d-23510286c013.png">

> > I think opening the docs (in the request view/area) and expanding the folder/group is an obvious and solid approach. It would also have the benefit of being easily expanded upon, allowing in the future to have Folder Plugins as buttons on the far right side of where requests have headers/authorization or other group/folder features (ie: authorization, headers, etc). Anyways, just like to ensure that I'm on the same page here.
> 
> My comment here, I feel like this should be a different issue, but I'm down to address it if you'd like, I personally really want some of those future features as being able to add authorization headers for ALL requests in a group would be AWESOME!

Yep these things would be separate issues for sure. Having workspace/folder defaults have been requested in the past. It's not in our internal roadmap yet (AFAIK) but if you want to solutions in respective issues I can get feedback from our designer about those. 😄

NickHackman committed:
fix: rename functions for handling showing the RequestGroupSettings modal

Co-authored-by: Opender Singh <opender94@gmail.com>

NickHackman committed:
fix(request-group-settings): remove incorrect rename functionality

Co-authored-by: Opender Singh <opender94@gmail.com>

NickHackman committed:
refactor: rename 'Request Group' in Settings to 'Folder'

NickHackman committed:
fix: remove RequestGroupMove as it is in Settings

No longer need Request Group Move as it is present in Settings. In addition, fix the implementation in Settings to be exactly the same as the implementation present in move-request-group-modal which is duplicate then remove current rather than solely an update.

NickHackman commented:
> We currently have a code-freeze in place because we're transitioning to typescript (1:1, should be fairly simple to update this PR), so I won't merge or test this just yet.

Awesome, Typescript sounds like a good choice for a project this big.

> Let's go with folder!

Done as of 9e642d1!

> That seems fine to me because we get it for free by deriving the implementation from `RequestSettingsModal`. It would be worth checking there is nothing special happening in the "Move" dropdown item here. If so, let's bring that over and remove this entry.

Removed and made necessary changes to have the same implementation as the modal that was used prior in 90a6aa5. Weirdly the previous implementation was a duplicate followed by a remove, not entirely sure how that's different than an update, but if it works, it works 😆 

> Yep these things would be separate issues for sure. Having workspace/folder defaults have been requested in the past. It's not in our internal roadmap yet (AFAIK) but if you want to solutions in respective issues I can get feedback from our designer about those. 😄

Good to hear, I'll have some up in the near future 🙂

develohpanda commented:
> Weirdly the previous implementation was a duplicate followed by a remove, not entirely sure how that's different than an update, but if it works, it works 😆

That checks out! The [`duplicate`](https://github.com/Kong/insomnia/blob/bdb4b4e6610f69e84a0ae7ea1296078c59bbe88e/packages/insomnia-app/app/common/database.js#L623-L669) and [`remove`](https://github.com/Kong/insomnia/blob/bdb4b4e6610f69e84a0ae7ea1296078c59bbe88e/packages/insomnia-app/app/common/database.js#L431-L449) db functions will operate on all children of the entity (`request-group` in this case). The [`update`](https://github.com/Kong/insomnia/blob/bdb4b4e6610f69e84a0ae7ea1296078c59bbe88e/packages/insomnia-app/app/common/database.js#L404-L429) will only operate on the entity itself.

NickHackman committed:
Merge branch 'develop' of https://github.com/Kong/insomnia into request-group-descriptions

NickHackman committed:
refactor(request-group-settings-modal): port to typescript

NickHackman commented:
@develohpanda fixed merge conflicts and ported new file to Typescript. Should be ready for review again :+1:

NickHackman commented:
@develohpanda any idea why the CI Windows Test is failing? Looks like a permission issue ([line](https://github.com/Kong/insomnia/pull/3350/checks?check_run_id=2862047093#step:5:617)) seems unrelated, but I don't know how others are passing and this has failed twice. Previous instance failed due to a connection reset ([line](https://github.com/Kong/insomnia/runs/2497518652#step:8:54)). 🤷‍♂️

develohpanda commented:
@NickHackman ah, windows has been a bit buggy recently and failing on random issues like that. I'll kick it off again!

develohpanda reviewed (approved):
LGTM, looks to be working well! We have some improvements we want to make in the future with modals (now that we are in TS) but until that's documented and the existing modals are updated, what you've done here is perfect. 🙌🏽 

<img width="983" alt="image" src="https://user-images.githubusercontent.com/4312346/122745787-748dfd80-d2dd-11eb-8ed3-d8aca933fe25.png">

develohpanda committed:
add line break

wdawson commented:
Looks great to me too. Thanks, @NickHackman for the PR! ⛵ 

Also, hi! I'm the new PM on the Insomnia team 👋

NickHackman commented:
Thanks for all your help @develohpanda, appreciate it and I'll for sure be contributing more in the future.

@wdawson That's awesome! I have some other issues open that I'd love your opinion on 😉 (#3368, #3351)

Looks like Windows is failing again due to connection reset, if you don't mind @develohpanda restarting the build again

develohpanda commented:
Just FYI @NickHackman waiting on one more eng review before merging 😄 almost good to go

NickHackman committed:
Merge branch 'develop' into request-group-descriptions

NickHackman committed:
Merge branch 'develop' into request-group-descriptions

dimitropoulos committed:
Merge branch 'develop' into request-group-descriptions

dimitropoulos committed:
removes errant `// @flow` comment

NickHackman commented:
@dimitropoulos whoops, missed the `@flow` annotation good catch!

dimitropoulos commented:
no worries @NickHackman! :)

Thanks for your contribution!

youssefsharief commented:
Is this in the current release?

develohpanda commented:
Hey @youssefsharief, this will be a part of the next release (which should be out this month)

youssefsharief commented:
@develohpanda thanks
